ios/Exponent/Versioned/Core/Api/EXNotifications.m

```diff
@@ -176,31 +182,137 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+        [newCategories removeObject:category];
+        break;
+      }
+    }
+    [newCategories addObject:newCategory];
+    [[EXUserNotificationCenter sharedInstance] setNotificationCategories:newCategories];
+    resolve(nil);
+  }];
+}
+
 #pragma mark - internal
 
-- (UILocalNotification *)_localNotificationFromPayload:(NSDictionary *)payload
+- (UNMutableNotificationContent *)_localNotificationFromPayload:(NSDictionary *)payload
 {
-  RCTAssert((payload[@"data"] != nil), @"Attempted to send a local notification with no `data` property.");
-  UILocalNotification *localNotification = [UILocalNotification new];
+  UNMutableNotificationContent *content = [UNMutableNotificationContent new];
 
   NSString *uniqueId = [[NSUUID new] UUIDString];
 
-  localNotification.alertTitle = payload[@"title"];
-  localNotification.alertBody = payload[@"body"];
-  
+  content.title = payload[@"title"];
+  content.body = payload[@"body"];
+
   if ([payload[@"sound"] boolValue]) {
-    localNotification.soundName = UILocalNotificationDefaultSoundName;
+    content.sound = [UNNotificationSound defaultSound];
+  }
+
+  if ([payload[@"count"] isKindOfClass:[NSNumber class]]) {
+    content.badge = (NSNumber *)payload[@"count"];
+  }
+
+  if ([payload[@"categoryId"] isKindOfClass:[NSString class]]) {
+    content.categoryIdentifier = [self scopedIdForIdentifier:payload[@"categoryId"]];
+  }
+
+  content.userInfo = @{
+                       @"body": payload[@"data"],
+                       @"experienceId": self.experienceId,
+                       @"id": uniqueId
+                       };
+
+  return content;
+}
+
+- (NSString *)scopedIdForIdentifier:(NSString *)identifier {
+  return [NSString stringWithFormat:@"%@:%@", self.experienceId, identifier];
+}
+
+- (UNCalendarNotificationTrigger *)notificationTriggerFor:(NSNumber * _Nullable)unixTime repeatingEvery:(NSString * _Nullable)timePeriod
+{
+  NSDateComponents *dateComponents = [self dateComponentsFrom:unixTime];
+  if (timePeriod) {
+    NSCalendarUnit timePeriodToRepeat = [RCTConvert NSCalendarUnit:timePeriod];
+    [self mutateDateComponents:dateComponents toRepeatEvery:timePeriodToRepeat];
+    return [UNCalendarNotificationTrigger triggerWithDateMatchingComponents:dateComponents repeats:YES];
+  }
+
+  return [UNCalendarNotificationTrigger triggerWithDateMatchingComponents:dateComponents repeats:NO];
+}
+
+- (NSDateComponents *)dateComponentsFrom:(NSNumber * _Nullable)unixTime {
+  static unsigned unitFlags = NSCalendarUnitSecond | NSCalendarUnitMinute | NSCalendarUnitHour | NSCalendarUnitDay | NSCalendarUnitMonth |  NSCalendarUnitYear;
+  NSDate *triggerDate = [RCTConvert NSDate:unixTime] ?: [NSDate new];
+  NSCalendar *calendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
+  return [calendar components:unitFlags fromDate:triggerDate];
+}
+
+- (void)mutateDateComponents:(NSDateComponents *)dateComponents toRepeatEvery:(NSCalendarUnit)timePeriod
```

Comment:
This code can't handle a use case when one wants to schedule a notification that will trigger eg. every minute, but starting next week (in such case it will start triggering this week).

Reply:
It's not possible to use a subclass of `UNCalendarNotificationTrigger`, adding a notification request with a custom trigger fails with
```
Error Domain=NSCocoaErrorDomain Code=4097 "connection to service named com.apple.usernotifications.usernotificationservice"
```

So, to sum up we can't replicate scheduling a notification that will repeat every minute starting 3 hours from now.

We could either remove support for `repeat` option on iOS or only in such case use `UILocalNotification` which is internally transformed to `UNNotificationRequest` by iOS (with a trigger of a non-public class `UNLegacyNotificationTrigger`).

What do you think? cc @ide @tsapeta

Reply:
I've added `ExpoNotifications.legacyScheduleLocalRepeatingNotification` exported native method which uses deprecated `UILocalNotification` to schedule a repeating notification with first fire date in the future. Using such configuration warns that this will be removed in next SDK release. The change has been neatly committed in https://github.com/expo/expo/pull/2316/commits/9fc0237ad1a2406af140fd794461374cb3790b43, before which I've implemented throwing in such case (`(time && repeat)`, https://github.com/expo/expo/pull/2316/commits/89068f90caf6d7684a8ebd678c0f46553faf9edd), so it's easily revertible.

Reply:
I think along with deprecation of (`time && repeat`) we should add ability to schedule a notification date-components-way, but let's do this in a separate PR.